File: backend/src/bond_monitor/domain/portfolio/redemption.py
"""Redemption and put-offer proceeds for portfolio positions."""

from __future__ import annotations

from bond_monitor.domain.portfolio.models import PortfolioPosition


def price_gain_total(position: PortfolioPosition) -> float:
    """Положительная разница «номинал − чистая цена покупки» × количество."""
    clean_at_purchase = position.purchase_clean_price_pct / 100.0 * position.face_value
    diff = position.face_value - clean_at_purchase
    return diff * position.bonds_count


def net_redemption_amount(
    position: PortfolioPosition,
    tax_rate: float,
    *,
    is_put: bool = False,
) -> float:
    """Сумма к получению при погашении/пут-оферте после НДФЛ на курсовую разницу."""
    if is_put:
        price_pct = position.offer_price_pct or 100.0
        redemption_per_bond = position.face_value * (price_pct / 100.0)
    else:
        redemption_per_bond = position.face_value
    gross = redemption_per_bond * position.bonds_count
    clean_at_purchase = position.purchase_clean_price_pct / 100.0 * position.face_value
    taxable_gain = max(0.0, (redemption_per_bond - clean_at_purchase) * position.bonds_count)
    tax = taxable_gain * tax_rate
    return gross - tax

File: backend/internal/domain/portfolio/redemption.go
package portfolio

// PriceGainTotal is positive (face - clean purchase) × bonds count.
func PriceGainTotal(position PortfolioPosition) float64 {
	cleanAtPurchase := position.PurchaseCleanPricePct / 100 * position.FaceValue
	diff := position.FaceValue - cleanAtPurchase
	return diff * float64(position.BondsCount())
}

// NetRedemptionAmount returns maturity/put proceeds after income tax on price gain.
func NetRedemptionAmount(position PortfolioPosition, taxRate float64, isPut bool) float64 {
	var redemptionPerBond float64
	if isPut {
		pricePct := 100.0
		if position.OfferPricePct != nil {
			pricePct = *position.OfferPricePct
		}
		redemptionPerBond = position.FaceValue * (pricePct / 100)
	} else {
		redemptionPerBond = position.FaceValue
	}
	gross := redemptionPerBond * float64(position.BondsCount())
	cleanAtPurchase := position.PurchaseCleanPricePct / 100 * position.FaceValue
	taxableGain := max(0, (redemptionPerBond-cleanAtPurchase)*float64(position.BondsCount()))
	return gross - taxableGain*taxRate
}

func max(a, b float64) float64 {
	if a > b {
		return a
	}
	return b
}
